### ISSUE TITLE
Recommended tweaks for further development

### ISSUE THREAD

**Issue by LuyeT** [13k/invokation#9]
Great work. All in all a very solid concept, model and framework. It's pretty much what I was seeking.

However, I would like to provide some feedback on some issues I had.

1. 
Most minor, but also most importantly, is the actual combos. For one thing, they seem to have been implemented several major patches ago and not touched since. Would be great to have some functionality of users submitting combos of their own to practice. As it stands now, a fair number of the combos have little or no application in the current meta, and the combos used currently are of course also subject to change.

2.
When training for combos, it gets rather tedious skilling invoker for each exercise. I believe it would be best if by default, invoker was pre-leveled, and optionally there was an option to turn this off or on.

3.
The arena doesn't provide any metric other than time of execution. I believe t would be really useful if there was a target amount of damage for a given combo to strive against. This could either be provided simply by having some hardcoded numbers for a preleveled invoker (as in point 2), or have this be dynamic by making a formula for calculating.

Hope this is of help! Keep up the great work!

**Comment by 13k** [comment]
First of all, thank you very much for taking the time to provide feedback.

> Most minor, but also most importantly, is the actual combos. For one thing, they seem to have been implemented several major patches ago and not touched since. Would be great to have some functionality of users submitting combos of their own to practice. As it stands now, a fair number of the combos have little or no application in the current meta, and the combos used currently are of course also subject to change.

User submitted content was the goal from the start, the built-in combos were a way to kickstart the game with a small set of data to work with. Regarding combos being outdated, I partially agree, and most of the reason is that I didn't create the combos myself, I researched several guides on the internet and compiled a small set of them.

If you check the [project's board](https://github.com/13k/invokation/projects/1) (I think this is public, if not, let me know), I'm planning to implement user-submitted combos, but this depends on implementing a bunch of backend infrastructure (namely, an external database service) that I didn't get to yet and also implementing a "combo editor" (which I'm planning to base on dota's in-game guide creator).

> When training for combos, it gets rather tedious skilling invoker for each exercise. I believe it would be best if by default, invoker was pre-leveled, and optionally there was an option to turn this off or on.

Recently I've implemented recommended skill builds (orbs and talents), which show on each combo's details, it should help with leveling the skills. The initial reason to leave it up to the player is that each combo shouldn't have a predefined skill build, and forcing builds would feel too brittle for more experienced players. On the other hand, I agree that it gets tedious, specially for newer players. The option to turn it on/off seems like a good compromise, I'll check how/if that fits in the current UI/flow and add it if possible.

> The arena doesn't provide any metric other than time of execution. I believe t would be really useful if there was a target amount of damage for a given combo to strive against. This could either be provided simply by having some hardcoded numbers for a preleveled invoker (as in point 2), or have this be dynamic by making a formula for calculating.

This is part of a larger feature which is measuring performance. I've thought about this a while ago and concluded that it is (possibly) a 3-part issue.

First part is to measure and show metrics. Currently there are 3 metrics:

1. time (stopwatch timer on the top right)
2. "multiplier" or spell/item usage count (`NNx` number above the hero HUD)
3. damage amount (shows at the end of the combo right below the multiplier). This is currently bugged (#2), but I'm working on a fix right now.

These metrics are already shown to the player and he/she can strive to improve from there.

I didn't come up with other metrics, and I'm not even sure we would want more.

Second part is to aggregate metrics and show historical progress.

I'm planning to [implement a points system](https://github.com/13k/invokation/projects/1#card-26313724) along with a [scoreboard UI](https://github.com/13k/invokation/projects/1#card-25031981) that stores and shows the metrics over time. The points system should be like "MMR for combos".

Third part, and possibly not needed, is to calculate/hardcode benchmarks that the player's performance can be compared against.

If I understood correctly, this is your suggestion, and [I thought about that before too](https://github.com/13k/invokation/projects/1#card-25772795). The card is in the Icebox column, which means it's an idea that should be investigated before implementing.

The reason I'm not sure it's worth implementing is because benchmarks are often artificial (what are the real ideal values?), hard to maintain as things change (imagine balance patches altering spell values all the time) and get outdated as overall performance improves. On top of that, if combos are contributed by users, they are responsible to calculate the benchmark values and keep it updated, which is even more of a burden on combo authors, but combo authoring should be kept as simple as possible.

> Hope this is of help! Keep up the great work!

Again, thanks a lot for testing and providing feedback!
